### DISCUSSION TITLE
RFC: Unicode & SVG path submissions

### DISCUSSION THREAD

**Discussion by chinchang** [cssbattle/forum#339] [RFC (Request for comments)]
## Problem statement
It is possible to submit solutions that are not necessarily created with just CSS, arguably. These lie in 2 categories - unicode & SVG paths. The competition aspect of the game should ideally be affected only though CSS solutions and not the other ones.

## Further context
Firstly, we need to understand why the non-CSS solutions are termed as "non-CSS" solutions.

1. **SVG paths** - SVG tags/images are banned in CSSBattle, because they are a different way to create visuals and it is possible to replicate the targets with just SVG and not writing a line of CSS. But SVG paths can be used for some CSS properties too. Though this is not strictly SVG in this scenario, it is not strictly CSS too, for 2 reasons: 1) Replicating the shapes/sizes in the target is an important aspect of the game which has a very different way to do when done with SVG paths. In fact, quite a few solutions are done by just giving some colors through CSS/HTML and rest all shapes with paths! That defeats the purpose of "CSS"battle. 2) Though SVG paths are used as values to CSS properties, we feel its more of an embedded syntax to use SVG in CSS. It's very similar to how you can embed CSS within HTML with style tag, JS within HTML with script tag or embed SVG inside HTML with SVG tag. Similarly, CSS also allows you to use an altogether different declarative language inside it.

2. **Unicode** - Unicode characters are rendered just like any other usual text character in HTML. This means they might render differently on different browsers. Moreover, they might render very differently on a headless browser (which we use to evaluate submissions). The difference in rendering of text (and hence unicode characters) makes it unpredictable - to play with and to learn from. Eg. See the [top solution on this open-target](https://cssbattle.dev/play/69). If you want to use these characters to replicate a target, the only solution is to use external tools that give you a better sense of the evaluation engine. Ideally, the CSSBattle platform should be self-sufficient for the game's basic component - replicating targets, without need of external tools. Plus, any player should be able to look at an open-target solution and see how it replicates the target - which isn't the case with unicodes since preview on browser is very different from the evaluation engine (for which unicode solutions are tailored).

## Motivation
We have gone through numerous feedback, forum threads, mails, DMs etc. Here are some guidelines we had for an ideal solution:
- We want the game to be fair to everyone. The rules should be simple enough and someone shouldn't need to debate if a solution is valid or not. (we understand this is tricky)
- A 100% matching solution in front of a player should preview 100% like the target...at least in most used modern browsers
- CSS should be the tool to be used to play the game
- SVG and unicode are surely interesting tools for a lot of players. They are a means to push the boundaries of the game! We want them to retain them in a way so as not to affect the general competitive aspect of the game.

## Proposed solution

The following solution is assuming we are able to reliably detect non-CSS submissions. We understand this is an ongoing effort, but we can do that.

- We'll have the game's general gameplay to be strictly CSS. i.e. Only if you submit CSS-only solution, would you show up on the leaderboard.
- We'll remove submission restrictions on unicode & SVG (paths). You'll be able to submit these non-CSS solutions.
- If you submit a non-CSS solution, you would be ranked on a "secondary" leaderboard - non-CSS leaderboard.
- We are calling it non-CSS because today it's SVG paths and unicode, but we could have something else there in the future.
- All players will by default see the CSS-only leaderboard - the "primary" leaderboard. This means, if you want to show up on the primary leaderboard, you'll need to submit a CSS-only solution.
- Any player would be able to toggle to the secondary leaderboard too - where players would be ranked on their SVG & unicode submissions. Something like: 
![image](https://user-images.githubusercontent.com/379918/214936120-b1cb83cf-0d83-423c-9cf5-1703f9bcd748.png)
- Global score/rankings & rewards would only be based on CSS-only submissions.
- One important decision pending here is - whether previous leaderboards will change or not? We haven't finalized that really, but most probably we won't change the past leaderboards - primarily because in the long run old leaderboards will become insignificant compared to new ones that come up. Though we are still open to hearing otherwise.
- Player's submissions history would be single stream - but each submission tagged:
![image](https://user-images.githubusercontent.com/379918/223085037-d35e840b-1956-458e-beaf-7e1b873fdf09.png)
-  For open targets, we'll be showing CSS-only submissions to start with. We'll get back to it again for any improvements.

### Detection mechanisms
**SVG**
- SVG tags or SVG path strings being used anywhere in the style attribute will be restricted

**Unicodes**
- Non-ASCII characters in the submission would be restricted.

## References
- https://github.com/cssbattle/forum/discussions/263#discussioncomment-2817594
- https://github.com/cssbattle/forum/discussions/335
- https://github.com/cssbattle/forum/discussions/148
- https://github.com/cssbattle/forum/discussions/6

----

We look forward to everyone's comments on the proposed solution. Do give a 👍 if you like the proposal.

**Comment by Esger** [comment]
I'd be totally happy with this new system but..
It should be applied to all the past battles as well to get a realistic idea about your standings.
And this 'filter' should be available in the stats page too. I might have some targets without a real css solution and I'd like to be able to detect those.

**Comment by Beowolve** [comment]
The idea of separate leader boards was suggested multiple times already and I think it is a good solution for the problems this topic generates. However I think some questions need a more detailed explanation.

1. There are other features expect **clip-path** that allow the syntax of a SVG path. Most importantly **offset:path**, but there is also **shape-outside**. What about that?
2. Nothing was said about the existing leader boards. Is the plan, to recalculate all existing submissions, or start with the split leader boards on the next battle? If so, what about solutions that are now **99.9%** and are still 🥇, or **var(--)** solutions?
3. If the ranking gets changed, what about also changing the way equals scores get handled like I proposed here: https://github.com/cssbattle/forum/discussions/263#discussioncomment-3101182
This would be even more important if a re submission of the old solutions will be required (I hope not!)
4. There is one more thing that makes me think. About 90% of the recent records are unicode based, but most don't have PRO, so many won't have access to there old CSS solution and might need to reinvent what they did. I keep notes of all my solutions, but most wont and might have a little problem with that change, at least at the beginning.
I suggest the following:
_You could keep the 5 submissions limit, but at least list the best CSS and best none CSS solution as well. Like a pinned solution._
5. What about all the medals:
If the leader boards change the medals need to be recalculated, but all the top players have worked hard to earn them. I guess none of us wants to lose them. So I suggest having two sections of medals for both CSS and none CSS, so you can't lose your hard earned medals.

**Reply by chinchang** [reply]
> There are other features expect clip-path that allow the syntax of a SVG path. Most importantly offset:path, but there is also shape-outside. What about that?

SVG path strings in general will result in the submission being tagged as non-CSS.

> Nothing was said about the existing leader boards. Is the plan, to recalculate all existing submissions, or start with the split leader boards on the next battle? If so, what about solutions that are now 99.9% and are still 🥇, or var(--) solutions?

Good question. We haven't finalized that really, but most probably we won't change the past leaderboards - primarily because in the long run old leaderboards will become insignificant compared to new ones that come up. Though we are still open to hearing otherwise.

> If the ranking gets changed, what about also changing the way equals scores get handled like I proposed here: https://github.com/cssbattle/forum/discussions/263#discussioncomment-3101182
This would be even more important if a re submission of the old solutions will be required (I hope not!)

As mentioned above, for now plan is to not change previous leaderboards. So skipping the further question.

> There is one more thing that makes me think. About 90% of the recent records are unicode based, but most don't have PRO, so many won't have access to there old CSS solution and might need to reinvent what they did. I keep notes of all my solutions, but most wont and might have a little problem with that change, at least at the beginning.
I suggest the following:
You could keep the 5 submissions limit, but at least list the best CSS and best none CSS solution as well. Like a pinned solution.

Good idea with the pinned best solution! If need be, this would be a nice way to go!
One question though - what is the 90% here? If you mean 90% of all targets have unicode records, that is not right...it is way less :)

> What about all the medals:
If the leader boards change the medals need to be recalculated, but all the top players have worked hard to earn them. I guess none of us wants to lose them. So I suggest having two sections of medals for both CSS and none CSS, so you can't lose your hard earned medals.

Again, depends on old leaderboards being changed. I'll say let us finalize on the leaderboards thing first and then take up these questions.


@Beowolve really appreciate all the questions n feedback here! 🙏🏼

**Reply by Beowolve** [reply]
I think we can't just skip the equal score/rank problem, because that will be more important for the pure CSS leader board. There will be more equal scores on that leader boards. The diversity of solutions unicode and clip path added to the game is significant. With pure CSS only, a lot of targets might get the same charcount/score in the end. Examples for this are #1, #3, #7, #9...
If the amount of possible solutions get reduced by a lot, It will be more common, that is just a fact.

Also what do you mean with, the past leader boards wont get changed? So no new records will be possible anymore for all the old targets? The battle leader boards are locked in already. This would go into the suggested season direction, then?

The 90% was just a pure guess on the changes in the activity timeline. #10 is CSS, everything else is unicode there.

**Reply by chinchang** [reply]
>If the amount of possible solutions get reduced by a lot, It will be more common, that is just a fact.

True. Not saying we don't want to tackle that problem. But just suggesting that we would rather tackle it separately. One step at a time.

About the leaderboards, we should have clarified about the different leaderboards. So here is what we are thinking:
Battle leaderboards - won't change. They are anyways locked after battle finishes, so they stay as it is.
Target leaderboards - they will update, for all 140 targets. And will remain open forever. Each target's leaderboard will start showing CSS-only ranking by default, with option to toggle to non-CSS rankings.
Global leaderboard - Because the target leaderboards change, so will the dependent global leaderboard too.

**Reply by Beowolve** [reply]
What is the plan if this changes get activated?
Do you plan to recalculate all previous submissions automatically and split them by the new rules into CSS/None CSS (how many submissions are there in the DB at the moment by the way?), or are we expected to resubmit?

**Reply by chinchang** [reply]
Auto-migration would happen - resubmission won't be required.

**Comment by JDMeijers** [comment]
Another idea to make a transition maybe more realistic:

There is a big list of targets now and I think CSSBattle has not come to an end of life yet. Maybe it's time to start a new kind of 'season'. Let's say there is 1 more battle in this 'season' so we have a nice 150 targets done. After that the leaderboard is saved and the season will close. People can always look back at this season's leaderboard and play the targets over again, but the leaderboard won't change in any way. 

A new season will start after this with new battles/targets and a/the new set of rules applied to them (in this case the block svg/unicode rules). In this case nobody will lose their hard earned medals and it's a fair start to all new players as well. In the future seasons can be shorter of course than 150 targets.

**Reply by Esger** [reply]
Closing a season means no more submissions are accepted?

**Comment by Beowolve** [comment]
If a season system gets implemented, how about making all old season targets open if a new season has started.
There would be a lot to learn for everyone then. I would have no problem with that.

**Reply by Esger** [reply]
In that case something should be done with the visible solutions, as only the top 10 (copied and mostly non-css) solutions are visible now.

**Comment by SelenIT** [comment]
Thanks for the great proposed solution!

I'd like to clarify: will the solution be considered Unicode only if it contains non-ASCII characters? It would leave some room for treating some solutions relying on text rendering as "pure CSS", but it's a clear and likely easy-to-implement criterion, so I'd be happy with it. Also, probably there should be no distinction between using Unicode characters "as is" and via character entities, as their rendering is identical and requires using extra tools for correct display?

**Reply by Beowolve** [reply]
Good question, I think it is totally fine how a simple dot is used in [#91 - reflection](https://cssbattle.dev/play/91). The rendering is equal in all browsers and the dot isn't used to render the output. Also the current unicode detection does not complain here, so it is most likely ok if the same detection is used in the future.
Of  course that would leave some characters free to use, basically everything you find on your keyboard.

Chars like
```
. _ | : ° +
```

**Reply by chinchang** [reply]
Ideally, any text should be restricted as they render differently on different systems. How about this - Instead of looking for non-ascii chars, we switch to detecting non-whitespace text nodes? Reason being - even if someone uses `<div>K</div>`, this will render differently on different systems. Of course, there might be ascii chars that render the same everywhere, but its the ones that don't render same that we need to be worried about.
With the above detection, of course whitespaces would still be allowed and we can't remove them for reason already mentioned before - we can't punish properly formatted submissions. So I guess we can go with the rule that "text nodes with non-whitespaces would not be allowed, except inside STYLE tags". 

There are stillI2 issues with this:
1.  a STYLE tag can also be rendered with a `display: block` style! So it could be used to put in text.
2. Plus there is `content: 'K'`. This we could still put a check for non-whitespace `content` value.

We did still like to discuss this further to come to a optimum rule. Until then, non-ascii check is what we have.
PS: Updated original RFC with detection mechanism for uncode (non-ASCII check for now) n svg.

**Reply by Beowolve** [reply]
What about the other "white space" characters then. They can be useful to change the way margins are calculated (to prevent collapsing margins) and won't be visible. At the moment you get a warning about them, but shouldn't the spaces be allowed then.

1. Is that what you mean? I like how you are thinking here, but if you check for unicode characters this would be detected now as well. And if anyone manages to make a record on a target with just regular characters and such a trick, it is well deserved ;)
   ```<style>*{_ █ __;display:block;font:299q'';margin:-397 0;color:b5e0ba;background:#5d3a3a```
   That is a 💯 match for [target #1](https://cssbattle.dev/play/1) by the way 😄 
2. `content:''` should be the only thing allowed if you don't want text to be used. Otherwise you could also use something like this to get around that limitation:
   ```<style>*::after{--x:"text";content:var(--x)```

**Reply by SelenIT** [reply]
Drawing the line between "text-based" and "non-text-based" solutions might be a bit more complicated than just detecting the text nodes outside `<style>`. On the one hand, there are

- list markers of `<li type=disc>` (default), `<li type=square>` etc. that are rendered as text characters and depend on the font rendering (I used it to get 238 chars in [Target #93](https://cssbattle.dev/play/93))
- the `quotes` property that allows to insert text as well that allows shorter solutions for targets like [#14](https://cssbattle.dev/play/14) (IIRC, @iamvdo prove it in the old forum)
- regular CSS units like `ex`, `ch`, and `ic` (with some more to come soon, like `cap`) that depend on the font used and thus render differently in different setups despite no actual text character used (my best solutions in [#8](https://cssbattle.dev/play/8) and [#81](https://cssbattle.dev/play/81) rely on them)

On the other hand, there are
- nearly-invisible (or just clipped-out) text characters with the only purpose to prevent margin collapsing or otherwise trigger the difference between empty and nonepty element, producing the same rendering anywhere (open solution of [#91](https://cssbattle.dev/play/91) being a good example)
- whitespace characters specifically defined in terms of their width (em-space, half-em-space, three-per-em-space etc.) and thus more-or-less guarranteed to produce the same rendering anywhere

Since there seems to be no reliable way to compare the built-in preview rendering with the backend rendering, I guess it would be better to account only the code itself, so solutions with font-relative units and, arguably, with list markers should be considered non-text-based... but it's not my strong opinion 😳

**Reply by SelenIT** [reply]
Upon reflection, looking back at the results of the [Battle #23](https://cssbattle.dev/battle/23), I have to admit that presence of non-ASCII was probably not the best criterion for telling CSS-solutions and text-based-solutions apart 😳. Despite non-ASCII chars provide way more options for building complex shapes, "ASCII art" still has enough simple shape primitives like dots/circles and bars to solve the targets like [#141](https://cssbattle.dev/play/141) in the same manner as "Unicode solutions" do — and with the similar score. It would seem more natural to me to treat such solutions the same way as "Unicode solutions" 😳

Unfortunately, I see no better criterion for "non-text-based" solutions now than 1) having no non-whitespace text content between tags (i.e. something like `solution.replace(/<style>.+?(<\/style>|$)/ig,'').replace(/<.+?>/g,'').replace(/[ \t\r\n\v]/g,'').length === 0`) and 2) containing no non-empty strings. Would it be too complicated to implement? 😳

**Reply by chinchang** [reply]
So if I understand correctly, all that will remain after these 2 checks would be:
1. list markers
2. strings in the style tag. Eg. `quotes: '"' '"' "'" "'";`

Correct @SelenIT?

**Reply by SelenIT** [reply]
I meant that the 2nd check is for all nonempty quoted strings inside `<style>` tags and `style` attributes, which can be found in `contents`, `quotes`, `list-style-type` and other CSS properties (including custom properties). There can be "false positives", e.g., with quoted font names in `font`, but fine-tuning `font` is likely a sign that the solution is going to use text anyway.

I have no strong opinion how to treat default list markers of certain elements (filled dot of `<li>`, triangle of `<details>` etc.), probably it would be fine to treat those as CSS 😳

**Reply by Beowolve** [reply]
> Upon reflection, looking back at the results of the [Battle #23](https://cssbattle.dev/battle/23), I have to admit that presence of non-ASCII was probably not the best criterion for telling CSS-solutions and text-based-solutions apart 😳. Despite non-ASCII chars provide way more options for building complex shapes, "ASCII art" still has enough simple shape primitives like dots/circles and bars to solve the targets like [#141](https://cssbattle.dev/play/141) in the same manner as "Unicode solutions" do — and with the similar score. It would seem more natural to me to treat such solutions the same way as "Unicode solutions" 😳
> 
> Unfortunately, I see no better criterion for "non-text-based" solutions now than 1) having no non-whitespace text content between tags (i.e. something like `solution.replace(/<style>.+?(<\/style>|$)/ig,'').replace(/<.+?>/g,'').replace(/[ \t\r\n\v]/g,'').length === 0`) and 2) containing no non-empty strings. Would it be too complicated to implement? 😳

If the regex is applied in this order, you will get problems with solutions that have a newline in it. To fix that, the order of the regex could be changed to remote all \t\r\n first, or change the first rule to:

`/<style>(?s:.)+?(<\/style>|$)/ig`

This should also include newline after `<style>`, or you will get false positive for regular solutions just because a newline is used, thast can't be desired of course.

**Comment by Beowolve** [comment]
After the solutions are split we will need updates on other parts of the webpage as well.
* The statistics/analytics page will need a filter for this (pure CSS / other solutions)
* The search option for previous solutions needs that filter as well
* The activity timeline will need more info. Something like "xxx broke new "pure CSS" record on target yyy". Maybe also split the timeline into both categories

**Comment by Beowolve** [comment]
I would like to see medals for both leader boards if this gets changed.
It could be a separate category listed below the regular medals. Unicode solutions usually require significantly more time to get working and  we have put a lot of effort into them in the past years so I feel it should still be rewarded with a medal. The way it is called in your preview screenshot frames it as the leader board for advanced player, and that is definitely true. Advanced solutions should get that as well in my opinion.

**Reply by Esger** [reply]
A leaderboard without medals is like a 99.9% solution :)

**Comment by Beowolve** [comment]
Here is an other example of a way to use Unicode without the current method recognizing it.

`<li><style>*{background:#5d3a3a;color:b5e0ba;margin:-17-24;list-style:'▀';font:4in'`

but it is basically the same as

`<li><style>*{background:#5d3a3a;color:b5e0ba;margin:65;list-style:square;scale:4.5`
or 
`<li type=square><style>*{background:#5d3a3a;color:b5e0ba;margin:65;scale:4.5`

And this should be considered regular CSS with no Unicode in my opinion.

Here is an other way to bypass the current Unicode tests.

`<body bgcolor=5d3a3a><style>html::before{color:b5e0ba;font:299q/40q'';content:open-quote;quotes:"\2588""`

**Comment by paperluigis** [comment]
about unicode, maybe just make both the preview and the headless browser use the same font?
though it would break the "html is passed as is" rule (unless the font is added dynamically with JS)

**Comment by Beowolve** [comment]
There is one thing, that hasn't been mentioned yet. What about the open targets and split leaderboards? I mean there is the possibility to show solutions from both leaderboards or only from the main leaderboard.
Personally I would like to see both.
Regardless of the open target solution, I think the submission history should also be split.

**Reply by chinchang** [reply]
Submissions history would be single - but each submission tagged:
![image](https://user-images.githubusercontent.com/379918/223085037-d35e840b-1956-458e-beaf-7e1b873fdf09.png)

For open targets, we'll be showing CSS only submissions to start with. We'll get back to it again for any improvements.

**Reply by chinchang** [reply]
Improved search functionality added to submission history -> https://discord.com/channels/909018468366549003/909018468936994848/1087339886480195674
